### DISCUSSION TITLE
DDNS remote access gives video stream

### DISCUSSION THREAD

**Discussion by tatsuo777github** [GerryDazoo/Slinger#442] [Q&A]
Gerry, thanks for the great work. You saved my three slingboxes that I have been using extensively. 

I setup my three slingboxs and the server on the same network and use DDNS.

When I type http://external IP address:port/Remote/BRX7020, I get the remote screen as I expect
When I type http://DDNS domain name:port/Remote/BRX7020, I get a video stream not the remote screen.
It is strange because these two ways of putting in the target device should give the same result.
Does anyone have any idea why?
Everything else works fine.

Here is my config.ini
[SLINGBOXES]
sb1=BRX7020
sb2=UBX7050
sb3=BRX6000A

[BRX7020]
sbtype="350/500/M1/M2"
password=E1:xx
ipaddress=192.168.1.8
port=5314
VideoSource=1
Remote=remote_panasonic_jisaku.txt
[UBX7050]
sbtype="Solo/Pro/ProHD"
password=E1:xx     
ipaddress=192.168.1.22
port=5301
VideoSource=1
Remote=remote_panasonic_jisaku.txt
[BRX6000A]
sbtype="350/500/M1"
password=E1:xx
ipaddress=192.168.1.17
port=5323
VideoSource=1
Remote=remote_panasonic_jisaku.txt
[SERVER]
port=36000
maxstreams=10
;enableremote=yes
URLbase=momo
;[REMOTE]
code=1

**Comment by GerryDazoo** [comment]
do a ping on your ddns address to make sure it's the same as your working Remote external address.
Please post the server output for both examples. 
PS. Delete the ;[Remote] and code=1 lines they have no effect.

**Reply by tatsuo777github** [reply]
Hi Gerry
Thanks for your quick reply.
Yes I did make sure by pinging my ddns address.
Here is what I get when I typed http://ddns/domain name:port/Remote/BRX7020

Version : 3.08f Running on Windows-10-10.0.19044-SP0 pid= 70644 C:\Slinger\Slinger-main\Slinger-main\slingbox_server.exe
Using config file config.ini
Connection Manager Running on port 36000 with 10 max streams using URL momo.
BOXES [('sb1', 'BRX7020'), ('sb2', 'UBX7050'), ('sb3', 'BRX6000A')]
Streamer Running:  10 config.ini UBX7050 UBX7050 36000 8388608
Streamer Running:  10 config.ini BRX7020 BRX7020 36000 8388608
Streamer Running:  10 config.ini BRX6000A BRX6000A 36000 8388608
Checking for slingbox at ('192.168.1.22', 5301)
starting up on  port 36000 8388608
Checking for slingbox at ('192.168.1.8', 5314)
Checking for slingbox at ('192.168.1.17', 5323)
('192.168.1.8', 5314) OK
BRX7020 Using slingbox at  ('192.168.1.8', 5314)
Streamer:  BRX7020 Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
('192.168.1.17', 5323) OK
('192.168.1.22', 5301) OK
BRX6000A Using slingbox at  ('192.168.1.17', 5323)
UBX7050 Using slingbox at  ('192.168.1.22', 5301)
Streamer:  BRX6000A Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
Streamer:  UBX7050 Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
Building page for BRX7020
Reading Custom Remote definition from remote_panasonic_jisaku.txt
Building page for UBX7050
Reading Custom Remote definition from remote_panasonic_jisaku.txt
Building page for BRX6000A
Reading Custom Remote definition from remote_panasonic_jisaku.txt
 * Serving Flask app 'slingbox_server' (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
12/18/2022, 22:47:28.148   Streaming request from ('45.8.68.43', 45872)
Starting remote stream 1

Slinginfo  "350/500/M1/M2" 12 30 192.168.1.8 5314 3072 10
BRX7020 Starting Stream for  45.8.68.43:45872
Connecting... ('192.168.1.8', 5314) Control
BRX7020 Using encrypted password: E1:xx
BRX7020 Selecting Video Source 1
Keycodes= [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 46, 47, 48, 50, 62, 63, 65, 66, 67, 68, 69, 70]
VideoParameters: Resolution= 12 FrameRate= 30 VideoBandwidth= 2000 VideoSmoothness= 63 IframeRate= 5 AudioBitRate= 64
Connecting... ('192.168.1.8', 5314) Stream
BRX7020 Stream started at 12/18/2022, 22:47:30.407  749 0
12/18/2022, 22:47:37.49  BRX7020 Stream Terminated for  45.8.68.43:45872
0 active remote connections
BRX7020 Shutting down connections
BRX7020 Logging Out
Streamer:  BRX7020 Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox

**Comment by tatsuo777github** [comment]
Minor correction:
Here is what I get when I typed http://ddns/domain name:port/Remote/BRX7020
-> Here is what I get when I typed http://host_name.domain_name:port/Remote/BRX7020

**Reply by GerryDazoo** [reply]
The code looks for the word "Remote" in the URL and forwards the request to the remote control handler. So it looks like for whatever reason it's not seeing that in the request although it sure looks like you typed it correctly.  When it fails, you see it downloading the video stream to your PC?
This would be easy to track down if you could run the python code OR if you know how to get a packet capture with Wireshark.
I have a similar configuration and it works OK for me and most everybody else.

**Reply by tatsuo777github** [reply]
>When it fails, you see it downloading the video stream to your PC?
Yes it downloads the video stream to my PC instead of the remote control screen.
I will look into running python or learn how to capture the packts.
I am extremely impressed with your prompt response. Thanks Gerry !

**Reply by mdehrlich** [reply]
Why is your enableremote=yes  commented out in your config file?

**Reply by GerryDazoo** [reply]
It's the default value.

**Reply by GerryDazoo** [reply]
If you get a chance try a different browser and/or delete anything the browser may have cached for your ddns domain.

**Reply by tatsuo777github** [reply]
Hi Gerry
enableremote=yes is not commented out in the config.ini I am running. I apologize for the confusion.
I tried different PC and smartphone with different browser but the results are same.

Then I ran slingbox_server.py. When I typed http://ddns.xxx.com:36000/Remote/BRX7200, .py gives video stream just as .exe did. However, when I typed http://192.168.1.20:36000/Remote/BRX7200, it gives different results from what I get from  slingbox_server.exe. 
The code stops without giving remote screen. 
The following is what I get. Is this normal ? 

Version : 3.08f Running on Windows-10-10.0.19044-SP0 pid= 44820 C:\Slinger\Slinger-main\Slinger-main\slingbox_server.py
Using config file config.ini
Connection Manager Running on port 36000 with 10 max streams using URL momo.
BOXES [('sb1', 'BRX7020'), ('sb2', 'UBX7050'), ('sb3', 'BRX6000A')]
Traceback (most recent call last):
  File "C:\Slinger\Slinger-main\Slinger-main\slingbox_server.py", line 1460, in <module>
Streamer Running:  10 config.ini BRX6000A BRX6000A 36000 8388608
Streamer Running:  10 config.ini UBX7050 UBX7050 36000 8388608
    from flask import Flask, render_template, request, render_template_string, abort, send_file
ModuleNotFoundError: No module named 'flask'
Streamer Running:  10 config.ini BRX7020 BRX7020 36000 8388608
starting up on  port 36000 8388608
Checking for slingbox at ('192.168.1.17', 5323)
Checking for slingbox at ('192.168.1.22', 5301)
Checking for slingbox at ('192.168.1.8', 5314)
('192.168.1.17', 5323) OK
('192.168.1.8', 5314) OK
BRX6000A Using slingbox at  ('192.168.1.17', 5323)
BRX7020 Using slingbox at  ('192.168.1.8', 5314)
Streamer:  BRX6000A Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
Streamer:  BRX7020 Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
('192.168.1.22', 5301) OK
UBX7050 Using slingbox at  ('192.168.1.22', 5301)
Streamer:  UBX7050 Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
12/20/2022, 16:58:04.032   RemoteControl connection from ('192.168.1.21', 54299)

Starting remote control stream handler for  ('192.168.1.21', 54299) to port 36001
Exception in thread Thread-5 (remote_control_stream):
Traceback (most recent call last):
  File "C:\Users\momoy\AppData\Local\Programs\Python\Python311\Lib\threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "C:\Users\momoy\AppData\Local\Programs\Python\Python311\Lib\threading.py", line 975, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Slinger\Slinger-main\Slinger-main\slingbox_server.py", line 1025, in remote_control_stream
    remote_control_socket.connect(('127.0.0.1', http_port ))  ## Send Packets to Flask
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ConnectionRefusedError: [WinError 10061] 対象のコンピューターによって拒否されたため、接続できませんでした。(this says "The connection could not be made because it was rejected by the target computer."

Tatsuo

**Reply by GerryDazoo** [reply]
You haven't properly configured your python.
from flask import Flask, render_template, request, render_template_string, abort, send_file
ModuleNotFoundError: No module named 'flask'

Please install the flask module. "ie. python3 -m pip install Flask.

And about your original problem. PLease try a different browser and/or delete any brower data that may be cached for your ddns domain name.

**Reply by surveyor2** [reply]
For me the remote doesn't work unless there is a stream already running. Is there a way to run the remote first without a stream running to (say) wake up the attached device so it doesn't confuse the stream server?

I also use ddns instead of external IP btw.

Jim

**Reply by GerryDazoo** [reply]
Jim: Set the "StartChannel" in your config.ini to send a wakeup keycode.
StartChannel=35+  Where 35 is the keycode to send to the device to wake it up. 35 is just an example, it will differ depending on what remote control in configured.

**Reply by tatsuo777github** [reply]
Hi Gerry
Thanks.
I have tried different browsers on different devices and the results are all the same. I get streaming not the remote screen.
Tatsuo

**Reply by GerryDazoo** [reply]
Did you get your Python running? Please post a screenshot of your browser when this is happening.

**Reply by tatsuo777github** [reply]
Gerry, yes I did run Python with flask installed. Now the result is consisitent with that of .exe 
Attached is the screenshot of server and browser screen

[server.txt](https://github.com/GerryDazoo/Slinger/files/10282311/server.txt)

![Browser_Screenshot](https://user-images.githubusercontent.com/120931332/209019005-35c43db2-2866-454d-a71f-920956fd8213.jpg)

Where should I go from here ?
Tatsuo

**Reply by surveyor2** [reply]
> Jim: Set the "StartChannel" in your config.ini to send a wakeup keycode. StartChannel=35+ Where 35 is the keycode to send to the device to wake it up. 35 is just an example, it will differ depending on what remote control in configured.

Thanks for the tip, Gerry. I'll play around with it a bit. It's a straight toggle on/off power button on the remote so I don't want to switch it off if it's already on.

Jim

**Reply by GerryDazoo** [reply]
Tatsuo: Firstly you can try using the complete correct URL
http://ddnsxxx:36000/Remote/BRX7020
I don't see the http:// in your screen shot even though that's what you said you were typing in a earlier post.
And add this debugging line after line 1120  ` data = connection.recv(1024).decode("utf-8")'
print('Received Request', data)

Make sure the indentation of the lines are the same. Then we can really see what you're browser is sending to the server. 
PS. For this test you don't have to have a stream running.

**Reply by tatsuo777github** [reply]
Hi Gerry
Yes I put in http:// but it disapears when it shows up in the browser cmd line.
I will put the debug code but I am starting to travel today and won't be back until mid January, so my ability to work on this is limited for a while. Thanks again Gerry and happy holidays!

**Reply by GerryDazoo** [reply]
Jim,
I think we're are using different terminology. Wakeup and power up are two different things. Many people have a similar issue but their remote has a "wakeup" button because their box has gone to "sleep" after a period of inactivity to save power not powered off. 
How did this work when you were using slingplayer?

**Reply by surveyor2** [reply]
Sorry I wasn't specific enough. If I called up slingplayer (with M1) and got confronted with a no video message (overlaid on the top of the video screen) I would press the power button on the remote and that would be it. I have only ever tried vlc with slinger and in the same situation the lack of a stream would cause vlc to dip out after a few seconds. I will then restart it and switch to a the browser remote to press the on button. It's not a big deal. I'm still learning.
Maybe there is a timeout setting somewhere in vlc to change to stop it dipping out automatically? If I come across something I'll post about it.

Jim

**Reply by GerryDazoo** [reply]
Dang, I hate VLC. A major amount of support requests come from people's issues with that code.
THe problem is VLC can't handle the new video stream when the box powers up. To workaround this, set the "loop one" option on VLC. Then it will automatically restart the stream when it drops.

**Reply by surveyor2** [reply]
Okay thanks again, Gerry.

Jim

**Reply by GerryDazoo** [reply]
Jim,

Some remotes don't implement all the keycodes
Keycode 1 toggles Power
Keycode 2 Power On
Keycode 3 PowerOff

Try sending keycode 2.
PS.  While I was testing this I found a bug in the code. I said to set StartChannel to 2+ 
The code needs at least two codes :-( Use StartChannel=2+2 for now. this will be fixed in the next release.

**Reply by surveyor2** [reply]
Hi Gerry,
Neither of my remotes use 2 or 3
I did a quick test of 1+2 to test the toggle on/off and it behaved exactly as you predicted earlier. It powered on/off okay but vlc didn't survive the power on and continue to stream. I had to restart it

Jim
